stats/opentelemetry/opentelemetry.go

```diff
@@ -0,0 +1,245 @@
+	"google.golang.org/grpc/internal"
+
+	"go.opentelemetry.io/otel/metric"
+)
+
+var logger = grpclog.Component("otel-plugin")
+
+var canonicalString = internal.CanonicalString.(func(codes.Code) string)
+
+var joinDialOptions = internal.JoinDialOptions.(func(...grpc.DialOption) grpc.DialOption)
+
+// EmptyMetrics represents no metrics. To start from a clean slate if the
+// intended effect is to pick a subset of metrics, use this and add onto it.
+var EmptyMetrics = Metrics{}
+
+// MetricName is a name of a metric.
+type MetricName string
+
+// Metrics is a set of metrics to record. Once created, Metrics is immutable.
+type Metrics struct {
+	// metrics are the set of metrics to initialize.
+	metrics map[MetricName]bool
+}
+
+// Add adds the metrics to the metrics set and returns a new copy with the
+// additional metrics.
+func (m *Metrics) Add(metrics ...MetricName) *Metrics {
+	newMetrics := make(map[MetricName]bool)
+	for metric := range m.metrics {
+		newMetrics[metric] = true
+	}
+
+	for _, metric := range metrics {
+		newMetrics[metric] = true
+	}
+	return &Metrics{
+		metrics: newMetrics,
+	}
+}
+
+// Remove removes the metrics from the metrics set and returns a new copy with
+// the metrics removed.
+func (m *Metrics) Remove(metrics ...MetricName) *Metrics {
+	newMetrics := make(map[MetricName]bool)
+	for metric := range m.metrics {
+		newMetrics[metric] = true
+	}
+
+	for _, metric := range metrics {
+		delete(newMetrics, metric)
+	}
+	return &Metrics{
+		metrics: newMetrics,
+	}
+}
+
+// MetricsOptions are the metrics options for OpenTelemetry instrumentation.
+type MetricsOptions struct {
+	// MeterProvider is the MeterProvider instance that will be used for access
+	// to Named Meter instances to instrument an application. To enable metrics
+	// collection, set a meter provider. If unset, no metrics will be recorded.
+	// Any implementation knobs (i.e. views, bounds) set in the passed in object
+	// take precedence over the API calls from the interface in this component
+	// (i.e. it will create default views for unset views).
+	MeterProvider metric.MeterProvider
+	// Metrics are the metrics to instrument. Will turn on the corresponding
+	// metric supported by the client and server instrumentation components if
+	// applicable.
+	Metrics Metrics
+	// TargetAttributeFilter is a callback that takes the target string of the
+	// channel and returns a bool representing whether to use target as a label
+	// value or use the string "other". If unset, will use the target string as
+	// is.
+	TargetAttributeFilter func(string) bool
+	// MethodAttributeFilter is a callback that takes the method string and
+	// returns a bool representing whether to use method as a label value or use
+	// the string "other". If unset, will use the method string as is. This is
+	// intended to be used only for generic methods (see
+	// grpc.UnknownServiceHandler), and not registered static methods (from
+	// generated service protos).
```

Comment:
Maybe something like `Set this to record the method name of RPCs handled by grpc.UnknownServiceHandler, but take care to limit the values allowed, as allowing too many will increase cardinality and could cause severe memory or performance problems.` ?

Reply:
I appended "Take care to limit the values allowed, as allowing too many will increase cardinality and could cause severe memory or performance problems." If you feel strongly about switching the whole comment, let me know, I felt like it was important to call out generic vs. static.

Reply:
This option is settable for both the server and client-side stats handler, right?

Is it only honored on server-side?  If so, that should be included.

And then the bit about "from generated service protos" is too specific, since you can register handlers without using protobuf.

Reply:
Alright I'll switch over to your comment.

Reply:
Wait no this is client and server side. Client side you also can want a knob for filtering arbitrary method names. Generated service protos applies to client side, server side is registered protos + registered handlers.

Reply:
> Client side you also can want a knob for filtering arbitrary method names. 

I thought that was what the call option was for?

Reply:
https://github.com/grpc/proposal/blob/master/A66-otel-stats.md calls for the flow to be something like - 
Is the method registered/known?
 -- If it's registered/known, record it
 -- If it's not, use the method filter to record it.

But, yeah, in your case, you can achieve this with the knob since the knob is available to the users. Your call

Reply:
As per offline discussion, we decided to go the route you mentioned server side (really no other option). On client side, generic methods also go through StaticMethodCallOption exposed to users (i.e. Invoke() or NewStream()).

Reply:
Wrote a new documentation comment.